### PR TITLE
Adds an ocean.

### PR DESCRIPTION
**Pull request by I-am-Erk**
#### Summary
Features "There's an ocean in the East (or wherever you set the json) now. NBD.  Feel free to add content for it."

### Purpose of change

We have all these boats and nowhere to use them.

### Describe the solution

Initially I wanted to add a "nocean", a freshwater ocean using lake settings, but @ralreegorganon, who added lakes in the first place, as like "noooo erk, do it properly". So I did. Now you get a full salt water ocean. There's no content for it, but it's hooked up for everything and supports custom map specials and things. It will probably need some fine tuning in that regard but it's a very fine start I believe 

This system works by rapidly but smoothly increasing the threshold for what can be considered a lake every OMT after we hit the threshold. It's random but takes 1-3 OM before you start seeing the nocean.

Credit to ehughsbaird for figuring out the solution for world gen lag. Tons of credit should go to ralreegorganon not just for programming the lakes code that this is standing on top of, but for guiding me in the general process of doing this while keeping his temper as I run roughshod over his baby.

#### Describe alternatives you've considered

~~I want to do this with salt water, but I also want to get it PRd for testing. Salt water should be straightforward, but not small, so this gets us started.~~

It was straightforward, and not small, but I did the salt water thing. Alternatives include half assing this with fresh water but, well.

### Testing

Description fixed since this:
image
now you can stare out into the ocean and contemplate mortality. Game improves 300%.

image
Or you can just dive in, and sink, and let the end take you.

image
It's peaceful, down here. Nobody has added any seaweed yet.

image
oceans are deeper than lakes. They get deep a little too fast, that's for another day.

---
known... bug?
image
this is salt water, if you're near an ocean and a lake forms, the game thinks it is an ocean sometimes. I don't have a great solution for this yet but also, it's hard to find and it only happens close to the ocean so I'm not sure it's a bug or just brackish water.

On exam I see that lake_shore is still making the ocean have fresh water on the edge. I will try to fix that but I have to hang xmas lights first.

---
Mapgen portion only: I've checked all the transitions from E, W, N, and S oceans just to make sure it looks natural. Haven't seen any artifacts, I think it's pretty good.
image
image
image
image
Have fun out there guys.

### Additional context
Another step towards #70056 

### Oh yeah

If you're interested in making content that would go in a saltwater ocean, now's a good time to start. It works exactly like lake specials, but with the word "ocean" instead.

### CODE REVIEW

**Comment by MNG-cataclysm** [comment]
> Is there a way to make the ocean into saltwater?

I think the best way would be to add some new water terrain types that give salt water instead of fresh, and then add ocean tiles that are composed of these saltwater terrain types. Connect the spawning of those to a predefined ocean type the same way we do with lakes, and that should work.

**Comment by John-Candlebury** [comment]
> It's pretty neat to set it to 1 om in each direction, which gives you a large island about 5x5 OM across with some really cool formations. 

Amazing, Cataclysm II

**Comment by Rimplestilt** [comment]
> Could we have the ocean roll over into another continent after traveling a great distance in a given direction?

Can we have a world size parameter along with the ability for X and Y chunk grids to eventually wrap around, and essentially simulate an entire globe? xD

**Comment by I-am-Erk** [comment]
> > Could we have the ocean roll over into another continent after traveling a great distance in a given direction?
> 
> Can we have a world size parameter along with the ability for X and Y chunk grids to eventually wrap around, and essentially simulate an entire globe? xD

Sure, you just have to code and add it :)

**Comment by I-am-Erk** [comment]
> > It's pretty neat to set it to 1 om in each direction, which gives you a large island about 5x5 OM across with some really cool formations.
> 
> Amazing, Cataclysm II

I think it might even work to set the numbers negative and get a tiny island, but I'm not sure.

**Comment by Standing-Storm** [comment]
> I think it might even work to set the numbers negative and get a tiny island, but I'm not sure.

If it does that would be a great challenge scenario

**Comment by fairyarmadillo** [comment]
I can't wait till we get beaches 😎

**Comment by MNG-cataclysm** [comment]
With this, should the Aircraft Carrier be moved out from inland lakes and to the ocean, where most aircraft carriers are?

**Comment by Vaskritaya** [comment]
> With this, should the Aircraft Carrier be moved out from inland lakes and to the ocean, where most aircraft carriers are? If it's along the coast, that can give some incentive for players to seek out the eastern edges of the map, and can become the first map special for the new oceans.

Adding to this, I was curious about transit between the ocean and the great lakes and if the giant lake outlined in #70056 would be a viable location, and I can pretty safely say the answer is no. Even the tiny USS America is too wide to go through the locks, so short of constructing an extremely tiny carrier just to field like 6 planes in a war against canada, a carrier on the lakes isn't plausible. Making it an ocean special is 100% the best move as far as realism is concerned.

**Comment by I-am-Erk** [comment]
Eventually the carrier should be permanently moved to the ocean. In the meantime we may want to duplicate an ocean version that is also globally unique.

File: data/json/furniture_and_terrain/terrain-liquids.json [hunk 1]
```diff
@@ -304,6 +304,20 @@
     "connects_to": "WATER",
     "examine_action": "water_source"
   },
+  {
+    "type": "terrain",
+    "id": "t_saltwater_cube",
+    "//": "for eventual use with water z levels.  Currently non-functional.",
```
Comment (Procyonae) [inline data/json/furniture_and_terrain/terrain-liquids.json:310]:
```suggestion

```
Pretty sure these comments are redundant now

**Comment by Procyonae** [comment]
How are you getting to the ocean quickly in testing and how viable would it be to add a starting scenario at the coast to make testing for future PRs quicker?

**Comment by Procyonae** [comment]
Also several peeps are complaining about getting dumped into areas of the new mapgen that they're not happy about (middle of crazy dense forest for instance) so I presume this will also leave players that have gone say straight east in their save stuck on a thin peninsula and have to retrace their explored 180x180 OMs to get back to new land, some way to make this less annoying before stable rolls round would be good

**Comment by I-am-Erk** [comment]
There's not really any way to prevent that, this can be disabled quickly in global settings by turning generate oceans off but people do need to read changelogs, I can't help them if they don't

**Comment by Procyonae** [comment]
Personally I don't think that's a reasonable expectation of stable players but fair enough

**Comment by I-am-Erk** [comment]
If it's a major issue, prior to stable we may be able to work out a solution

**Comment by fairyarmadillo** [comment]
> Personally I don't think that's a reasonable expectation of stable players but fair enough

Would it be possible to reorient legacy saves so that their new coords place 0, 0, 0 at the player's location the first time they load?

**Comment by I-am-Erk** [comment]
That sounds like it would have serious repercussions. I'd prefer a script that sets the ocean edge value to be equal to the furthest they'd travelled, but that's a lot of work. Or the global setting for oceans could be turned off by default in older games, I'm not sure how to do that but it might be possible.

**Comment by I-am-Erk** [comment]
Hyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyype

**Comment by Zireael07** [comment]
Well deserved hype I'd say unlike some games...

**Pull request merged**